specs/agents/tracing-instrumentation-aws.md

```diff
@@ -43,17 +43,27 @@ The format should be `DynamoDB <ActionName> <TableName>`. So for example, `Dynam
 
 ### SQS (Simple Queue Service)
 
-AWS Simple Queue Service is a message queuing service. The [messaging spec](tracing-instrumentation-messaging.md) can 
+AWS Simple Queue Service is a message queuing service. The [messaging spec](tracing-instrumentation-messaging.md) can
 be used for instrumenting SQS, but the follow specifications supersede those of the messaging spec.
 
 - **`context.destination.cloud.region`**: mandatory. The AWS region where the queue is.
 
+#### Distributed Tracing
+
 For distributed tracing, the SQS API has "message attributes" that can be used in lieu of headers.
 
+Agents should use an attribute name of `Traceparent` or `Elastic-Trace-Parent` when sending the traceparent headers via the SQS message attributes.  Agents should use an attribute name of `Tracecontext` or `Elastic-Trace-Context` if sending tracecontext headers in an SQS message attribute.
+
+Agents should default to using the non-namespaced `Traceparent` and `Tracecontext` versions of the attribute names.  Agents should use the namespaced versions of the attribute names only if the `use_elastic_traceparent_header` configuration value is set to `true`.
+
+If agents attempt to automatically propagate headers when receiving messages they should check first for the non-namespaced versions, and then check for the namespaced version.  If both versions are present, agents should prefer the non-namespaced versions and ignore the namespaced versions.
+
+SQS has a documented limit of ten message attributes per message.  Agents _should not_ add traceparent or tracecontext headers to the message attributes if adding those fields would put an individual message over this limit.
```

Comment:
You sure we want to add a `.length` check to every single message's attributes? Does AWS' docs say anything about responsibilities? I'm not as worried about performance than I am about confusing behaviour for users. _"Why is it not adding the headers?"_

Reply:
One's never sure in engineering @mikker ;)  But here's my thinking on that.  In my testing, if the number of Message Attributes exceeds 10, then Amazon SQS will accept the request, the service API will return a result, but that result will look like this

```
{
  ResponseMetadata: { RequestId: '9cef7bc1-2e52-52fd-8ced-231d5a0cd243' },
  Successful: [],
  Failed: [
    {
      Id: 'foo',
      SenderFault: true,
      Code: 'InvalidParameterValue',
      Message: 'Number of message attributes [12] exceeds the allowed maximum [10].'
    }
  ]
}
```

Also, the message will not be enqueued.  This means if an agent adds an 11th and 12th (if you're doing tracestate) item to the message attributes, then the agent will stop the message from being enqueued.  My biases are usually towards preserving an end user's application behavior.

@mikker Do you have a feel for how expensive the length check would be in ruby?

Reply:
I demand absolute certainty 😅

I don't worry about the overhead really. I wouldn't worry we'll make an actual impact in a "real" app with a simple length check.

I'm more concerned about confusing users. I guess we could log a warning if we omit the headers due to 10 already being set? That way confused users looking for their headers could hopefully figure out the explanation quite easily.

Reply:
I agree that the traceparent not being added might catch users off guard -- but not as much as a sub-section of their  own queue messages not being sent when they deploy the agent. A log message sounds like a great compromise.  I will update the spec PR accordingly.

Reply:
Latest version: https://github.com/elastic/apm/blob/fb5686893f2c6683d7e79950f72c4764dfc4052c/specs/agents/tracing-instrumentation-aws.md#distributed-tracing